File: addons/guide-cpp/src/inputs/guide_input_mouse_axis_1d.cpp
#include "guide_input_mouse_axis_1d.h"
#include "guide_input_state.h"

using namespace godot;

void GUIDEInputMouseAxis1D::_bind_methods() {
    BIND_ENUM_CONSTANT(AXIS_X);
    BIND_ENUM_CONSTANT(AXIS_Y);

    ClassDB::bind_method(D_METHOD("get_axis"), &GUIDEInputMouseAxis1D::get_axis);
    ClassDB::bind_method(D_METHOD("set_axis", "axis"), &GUIDEInputMouseAxis1D::set_axis);
    ADD_PROPERTY(PropertyInfo(Variant::INT, "axis", PROPERTY_HINT_ENUM, "X,Y"), "set_axis", "get_axis");

    ClassDB::bind_method(D_METHOD("_refresh"), &GUIDEInputMouseAxis1D::_refresh);
    ClassDB::bind_method(D_METHOD("_to_string"), &GUIDEInputMouseAxis1D::_to_string);
}

GUIDEInputMouseAxis1D::GUIDEInputMouseAxis1D() {
}

GUIDEInputMouseAxis1D::~GUIDEInputMouseAxis1D() {
}

bool GUIDEInputMouseAxis1D::_needs_reset() const {
    return true;
}

void GUIDEInputMouseAxis1D::_begin_usage() {
    if (_state) {
        _state->connect("mouse_position_changed", Callable(this, "_refresh"));
        _refresh();
    }
}

void GUIDEInputMouseAxis1D::_end_usage() {
    if (_state) {
        _state->disconnect("mouse_position_changed", Callable(this, "_refresh"));
    }
}

void GUIDEInputMouseAxis1D::_refresh() {
    if (_state) {
        Vector2 delta = _state->get_mouse_delta_since_last_frame();
        if (axis == AXIS_X) {
            _value.x = delta.x;
        } else {
            _value.x = delta.y;
        }
    }
}

bool GUIDEInputMouseAxis1D::is_same_as(const Ref<GUIDEInput> &other) const {
    Ref<GUIDEInputMouseAxis1D> o = other;
    if (o.is_null()) return false;
    return o->axis == axis;
}

String GUIDEInputMouseAxis1D::_to_string() const {
    return "(GUIDEInputMouseAxis1D: axis=" + String::num(axis) + ")";
}

String GUIDEInputMouseAxis1D::_editor_name() const {
    return "Mouse Axis 1D";
}

String GUIDEInputMouseAxis1D::_editor_description() const {
    return "Relative mouse movement on a single axis.";
}

GUIDEAction::GUIDEActionValueType GUIDEInputMouseAxis1D::_native_value_type() const {
    return GUIDEAction::AXIS_1D;
}

GUIDEInput::DeviceType GUIDEInputMouseAxis1D::_device_type() const {
    return MOUSE;
}

File: addons/GuideCS/Inputs/GUIDEInputMouseAxis1D.cs
using System;
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEInputMouseAxis1D : GUIDEInput {
	public enum GuideInputMouseAxis {
		X,
		Y
	}

	private GuideInputMouseAxis _axis;
	[Export]
	public GuideInputMouseAxis Axis {
		get => _axis;
		set {
			if (_axis == value) return;
			_axis = value;
			EmitChanged();
		}
	}

	public override bool _NeedsReset() => true;

	public override void _BeginUsage() {
		_State.MousePositionChanged += _Refresh;
		_Refresh();
	}

	public override void _EndUsage() {
		_State.MousePositionChanged -= _Refresh;
	}

	private void _Refresh() {
		Vector2 delta = _State.GetMouseDeltaSinceLastFrame();
		switch (Axis) {
			case GuideInputMouseAxis.X:
				_Value = new Vector3(delta.X, 0, 0);
				break;
			case GuideInputMouseAxis.Y:
				_Value = new Vector3(delta.Y, 0, 0);
				break;
		}
	}

	public override bool IsSameAs(GUIDEInput other) {
		return other is GUIDEInputMouseAxis1D mouseAxis && mouseAxis.Axis == Axis;
	}

	public override string ToString() => $"(GuideInputMouseAxis1D: axis={Axis})";

	public override string _EditorName() => "Mouse Axis 1D";

	public override string _EditorDescription() => "Relative mouse movement on a single axis.";

	public override GUIDEAction.GUIDEActionValueType _NativeValueType() => GUIDEAction.GUIDEActionValueType.Axis1D;

	public override DeviceType _DeviceType() => DeviceType.Mouse;
}
